Java_observer/Observer_patten.java

```diff
@@ -0,0 +1,73 @@
+import java.util.ArrayList;
+import java.util.List;
+
+// Subject interface which store list
+interface StockMarket {
+    void addObserver(Investor investor);
+    void removeObserver(Investor investor);
+    void notifyObservers();
+}
+
+// Observer interface
+interface Investor {
+    void update(String stock, double price);
+}
+
+// Concrete Subject
+class StockMarketImpl implements StockMarket {
+    private List<Investor> investors = new ArrayList<>();
+    private String stock;
+    private double price;
+
+    @Override
+    public void addObserver(Investor investor) {
+        investors.add(investor);
+    }
+
+    @Override
+    public void removeObserver(Investor investor) {
+        investors.remove(investor);
+    }
+
+    @Override
+    public void notifyObservers() {
+        for (Investor investor : investors) {
+            investor.update(stock, price);
+        }
+    }
+
+    public void setStockPrice(String stock, double price) {
+        this.stock = stock;
+        this.price = price;
+        notifyObservers();
+    }
+}
+
+// Concrete Observer
+class InvestorImpl implements Investor {
+    private String name;
+
+    public InvestorImpl(String name) {
+        this.name = name;
+    }
+
+    @Override
+    public void update(String stock, double price) {
+        System.out.println(name + " received update - " + stock + " price: " + price);
+    }
+}
+
+//Main Class(Usage example)
+public class Observer_patten  {
+    public static void main(String[] args) {
+        StockMarketImpl stockMarket = new StockMarketImpl();
+
+        Investor investor1 = new InvestorImpl("Alice");
+        Investor investor2 = new InvestorImpl("Bob");
```

Comment:
I noticed here you have
`StockMarketImpl stockMarket = new StockMarketImpl();`
but
`Investor investor = new InvestorImpl("name");`
instead of
`InvestorImpl investor = new InvestorImpl("name");`
I'm just wondering if there is a reason on why you called the classes differently?

Also for this example, would it be better to make the subject a single `stock` instead of the whole market to showcase that an `observer` can `subscribe` to multiple `subjects` or in this case `stocks`?

Reply:
Naming Convention Consistency: The naming discrepancy you've noticed is likely an oversight. It's best practice to maintain consistency in naming conventions. In the provided code, both StockMarketImpl and InvestorImpl should be called with their full names when instantiated.

So, instead of:
Investor investor = new InvestorImpl("name");
It should be:
InvestorImpl investor = new InvestorImpl("name");